### PR TITLE
Add DeepEquals generated code

### PR DESCRIPTION
**Pull request by aanm**
Not relying on reflect.DeepEquals improves the DeepEqual performance
logic a lot. As an example, a benchmark was written for SpecEquals and
this commit was able to reduce the number of allocations and bytes per
operation to 0.

```
BenchmarkSpecEquals/Reflected_SpecEquals-8     98300  12325 ns/op  3117 B/op  68 allocs/op
BenchmarkSpecEquals/Generated_SpecEquals-8   3570753    354 ns/op     0 B/op   0 allocs/op
```

Note for reviewers: I had this sitting on my side for a while, I didn't submitted because of k8s structures but since we now control the k8s structures we can use these generated methods in our advantage.

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by tklauser — APPROVED** [review]

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
retest-runtime

**Comment by nebril** [comment]
retest-runtime

File: Makefile [hunk 1]
```diff
@@ -68,16 +68,37 @@ define generate_k8s_api
...
+	$(call generate_deepequal,"$(subst $(space),$(comma),$(foreach pkg,$(2),$(1)/$(subst ",,$(subst :,/,$(pkg)))))")
 endef
 
-define generate_k8s_api_deepcopy
+empty:=
+space:= $(empty) $(empty)
+
+define generate_k8s_api_deepcopy_deepequal
 	$(call generate_k8s_api,deepcopy,github.com/cilium/cilium/pkg/k8s/client,$(1),$(2))
+	# Explanation for the 'subst' below:
```
Comment (nebril) [inline Makefile]: Thanks for the comment!

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/cidr/cidr.go [hunk 2]
```diff
@@ -33,6 +33,25 @@ type CIDR struct {
 	*net.IPNet
 }
 
+func (in *CIDR) DeepEqual(other *CIDR) bool {
+	if other == nil {
```
Comment (nebril) [inline pkg/cidr/cidr.go:40]: Shouldn't we also check for `in` being nil here (Like in `CiliumNetworkPolicy` below)?

File: pkg/policy/api/selector.go [hunk 3]
```diff
@@ -47,6 +50,20 @@ type EndpointSelector struct {
...
 }
 
+func (in *EndpointSelector) DeepEqual(other *EndpointSelector) bool {
+	switch {
+	case (in == nil) != (other == nil):
+		return false
+	case (in == nil) && (other == nil):
+		return true
+	}
+	// Once k8s upstream library supports DeepEqual we can remove this reflect
```
Comment (nebril) [inline pkg/policy/api/selector.go]: Would be nice to have a gh issue tracking that.

**Review by nebril — COMMENTED** [review]
Good stuff! One question inline, one GH issue to create.

File: pkg/cidr/cidr.go (hunk 2 above)
Comment (aanm) [reply pkg/cidr/cidr.go:40]: It's up to the caller to check for the receiver to be nil.

**Comment by aanm** [comment]
test-me-please

**Review by nebril — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by vadorovsky — APPROVED** [review]

**Comment by aanm** [comment]
retest-net-next

**Review by christarazi — APPROVED** [review]
Nice!

**Comment by aanm** [comment]
hit https://github.com/cilium/cilium/issues/11181

**Pull request merged**